heat/cluster/spectral.py

```diff
@@ -0,0 +1,309 @@
+        self._labels = self._cluster.labels_
+        self._cluster_centers = self._cluster.cluster_centers_
+
+        return self
+
+    def predict(self, X):
+        """
+        Predict the closest cluster each sample in X belongs to.
+
+        X is transformed to the low-dim representation by calculation of eigenspectrum (eigenvalues and eigenvectors) of the graph laplacian from the similarity matrix.
+        Inference of lables is done by extraction of the closest centroid of the n_clusters eigenvectors from the previously fitted clustering algorithm (kmeans)
+        Caution: Calculation of the low-dim representation requires some time!
+
+        Parameters
+        ----------
+        X : ht.DNDarray, shape = [n_samples, n_features]
+            New data to predict.
+
+        Returns
+        -------
+        labels : ht.DNDarray, shape = [n_samples,]
+            Index of the cluster each sample belongs to.
+        """
+        # input sanitation
+        if not isinstance(X, ht.DNDarray):
+            raise ValueError("input needs to be a ht.DNDarray, but was {}".format(type(X)))
+        if X.split is not None and X.split != 0:
+            raise NotImplementedError("Not implemented for other splitting-axes")
+        # 1. Calculation of Adjacency Matrix
+        if self.metric == "rbf":
+            sig = math.sqrt(1 / (2 * self.gamma))
+            S = ht.spatial.rbf(X, sigma=sig, quadratic_expansion=True)
+
+        elif self.metric == "euclidean":
+            S = ht.spatial.cdist(X)
+        else:
+            raise NotImplementedError("Other kernels currently not implemented")
+
+        # 2. Calculation of Laplacian
+        if self.laplacian == "eNeighborhood":
+            if self.epsilon[1] == "upper":
+                L = laplacian(S, norm=self.normalize, mode="eNeighbour", upper=self.epsilon[0])
+            elif self.epsilon[1] == "lower":
+                L = laplacian(S, norm=self.normalize, mode="eNeighbour", lower=self.epsilon[0])
+            else:
+                raise ValueError(
+                    "Boundary needs to be 'upper' or 'lower' and threshold needs to be set, if laplacian = eNeighborhood"
+                )
+
+        elif self.laplacian == "fully_connected":
+            L = laplacian(self._similarity, norm=self.normalize, mode="fc")
+        else:
+            raise NotImplementedError("Other approaches currently not implemented")
+
+        # 3. Eigenvalue and -vector calculation via Lanczos Algorithm
+        v0 = ht.ones((L.shape[0],), dtype=L.dtype, split=L.split, device=L.device) / math.sqrt(
+            L.shape[0]
+        )
+
+        V, T = ht.lanczos(L, self.n_lanczos, v0)
+        # 4. Calculate and Sort Eigenvalues and Eigenvectors of tridiagonal matrix T
+        eval, evec = torch.eig(T._DNDarray__array, eigenvectors=True)
+        # If x is an Eigenvector of T, then y = V@x is the corresponding Eigenvector of L
+        eval, idx = torch.sort(eval[:, 0], dim=0)
```

Comment:
ht.sort?

Reply:
doesn't work, indexing with the returned ht.dndarray throws errors